### ISSUE TITLE
imwrite problems with ImageCmap

### ISSUE THREAD

**Issue by lobingera** [timholy/Images.jl#336]
Following a discussion in julia-users
https://groups.google.com/forum/#!msg/julia-users/QIGnmVWA-wE/JesKN6OoAgAJ

Images.jl (current 0.4.43) together with Version 0.4.0-dev+6738  Commit 181c19f* (0 days old master)
create in running this:
```
using Images, Color


datafloat = reshape(linspace(0.5, 1.5, 60000), 200, 300);

dataint = round(Uint8, 254*(datafloat - 0.5) + 1);

b = RGB(0,0,1)
w = RGB(1,1,1)
r = RGB(1,0,0)
cmaprgb = Array(RGB, 255)
f = linspace(0,1,128)
cmaprgb[1:128] = [(1-x)*b + x*w for x in f]
cmaprgb[129:end] = [(1-x)*w + x*r for x in f[2:end]]

img = ImageCmap(dataint, cmaprgb);

imwrite(img,"image.pbm")
```

This error message:
```
lobi@orange4:~/juliarepo$ ../julia04/julia 
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+6738 (2015-08-15 04:08 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 181c19f* (0 days old master)
|__/                   |  x86_64-linux-gnu

julia> include("test_image_cmap.jl")
ERROR: LoadError: MethodError: `imwrite` has no method matching imwrite(::Images.ImageCmap{Color.RGB{T<:Union{AbstractFloat,FixedPointNumbers.FixedPoint}},2,Array{UInt8,2}}, ::IOStream, ::Type{Images.PBMBinary})
Closest candidates are:
  imwrite{T<:Images.ImageFileType}(::Any, ::AbstractString, ::Type{T<:Images.ImageFileType})
  imwrite{T<:Color.ColorValue{T}}(::AbstractArray{T<:Color.ColorValue{T},N}, ::IO, ::Type{Images.PPMBinary})
  imwrite{T<:Color.ColorValue{T}}(::AbstractArray{T<:Color.ColorValue{T},N}, ::IO, ::Type{Images.PPMBinary}, ::Any)
  ...
 in anonymous at /home/lobi/.julia/v0.4/Images/src/io.jl:197
 in open at iostream.jl:114
 in imwrite at /home/lobi/.julia/v0.4/Images/src/io.jl:196
 in imwrite at /home/lobi/.julia/v0.4/Images/src/io.jl:186
 in include at ./boot.jl:254
 in include_from_node1 at ./loading.jl:264
while loading /home/lobi/juliarepo/test_image_cmap.jl, in expression starting on line 18

```

**Comment by lobingera** [comment]
Maybe something else, but...
```
lobi@orange4:~/juliarepo$ ../julia04/julia 
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+6738 (2015-08-15 04:08 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 181c19f* (1 day old master)
|__/                   |  x86_64-linux-gnu

julia> include("test_image_cmap.jl")
INFO: Recompiling stale cache file /home/lobi/.julia/lib/v0.4/Images.ji for module Images.

signal (11): Segmentation fault
unknown function (ip: 0x7f2e0dfec1f7)
unknown function (ip: 0x7f2e0dfeee85)
jl_restore_incremental at /home/lobi/julia04/usr/bin/../lib/libjulia.so (unknown line)
_require_from_serialized at loading.jl:83
jlcall__require_from_serialized_21134 at  (unknown line)
jl_apply_generic at /home/lobi/julia04/usr/bin/../lib/libjulia.so (unknown line)
_require_from_serialized at ./loading.jl:104
require at ./loading.jl:179
unknown function (ip: 0x7f2e0ac2ef8c)
jl_apply_generic at /home/lobi/julia04/usr/bin/../lib/libjulia.so (unknown line)
unknown function (ip: 0x7f2e0dff5b85)
unknown function (ip: 0x7f2e0dff6abb)
unknown function (ip: 0x7f2e0dff68cd)
unknown function (ip: 0x7f2e0dff74c9)
jl_load at /home/lobi/julia04/usr/bin/../lib/libjulia.so (unknown line)
include at ./boot.jl:254
jl_apply_generic at /home/lobi/julia04/usr/bin/../lib/libjulia.so (unknown line)
include_from_node1 at ./loading.jl:264
jl_apply_generic at /home/lobi/julia04/usr/bin/../lib/libjulia.so (unknown line)
unknown function (ip: 0x7f2e0dfe2d83)
unknown function (ip: 0x7f2e0dfe21a1)
unknown function (ip: 0x7f2e0dff6848)
jl_toplevel_eval_in at /home/lobi/julia04/usr/bin/../lib/libjulia.so (unknown line)
eval_user_input at REPL.jl:63
jlcall_eval_user_input_21120 at  (unknown line)
jl_apply_generic at /home/lobi/julia04/usr/bin/../lib/libjulia.so (unknown line)
anonymous at task.jl:92
unknown function (ip: 0x7f2e0dfe8950)
unknown function (ip: (nil))
Segmentation fault (core dumped)
```

in long:

```
lobi@orange4:~/juliarepo$ ../julia04/julia 
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+6738 (2015-08-15 04:08 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 181c19f* (1 day old master)
|__/                   |  x86_64-linux-gnu

julia> Pkg.status()
WARNING: unknown Clang commit c4d18994, metadata may be ahead of package cache
14 required packages:
 - BinDeps                       0.3.14
 - Clang                         0.0.5+             9b668ecf
 - Dates                         0.4.4
 - Gadfly                        0.3.13
 - GraphViz                      0.0.4              master
 - Graphs                        0.5.5
 - Gtk                           0.8.2+             new_canvas
 - ImageView                     0.1.15
 - Images                        0.4.43+            master
 - PyCall                        0.8.2
 - PyPlot                        1.5.3
 - Tk                            0.3.4+             f283145f
 - Vega                          0.3.3
 - Winston                       0.11.11
46 additional packages:
 - ArrayViews                    0.6.2
 - Cairo                         0.2.28+            canvas_new (dirty)
 - Calculus                      0.1.10
 - Codecs                        0.1.4
 - Color                         0.4.7
 - ColorBrewer                   0.1.1
 - Compat                        0.5.1
 - Compose                       0.3.13+            text_scale
 - Contour                       0.0.7
 - Cxx                           0.0.0-             master (unregistered)
 - DataArrays                    0.2.17
 - DataFrames                    0.6.8
 - DataStructures                0.3.11
 - Distances                     0.2.0
 - Distributions                 0.8.3
 - Docile                        0.5.16
 - DualNumbers                   0.1.3
 - FixedPointNumbers             0.0.10
 - GZip                          0.2.17
 - Graphics                      0.1.1
 - Grid                          0.3.10
 - Hexagons                      0.0.3
 - ImmutableArrays               0.0.9
 - IniFile                       0.2.4
 - Iplot                         0.0.0-             master (unregistered, dirty)
 - Iterators                     0.1.8
 - JSON                          0.4.4
 - KernelDensity                 0.1.1
 - LaTeXStrings                  0.1.4
 - Loess                         0.0.3
 - NaNMath                       0.0.2
 - Optim                         0.4.2
 - PDMats                        0.3.5
 - Reexport                      0.0.2
 - Rsvg                          0.0.0-             master (unregistered, dirty)
 - SHA                           0.1.0
 - SIUnits                       0.0.5
 - Showoff                       0.0.4
 - SortingAlgorithms             0.0.5
 - StatsBase                     0.7.0
 - StatsFuns                     0.1.1
 - TexExtensions                 0.0.2
 - URIParser                     0.0.5

julia> include("test_image_cmap.jl")
ERROR: LoadError: MethodError: `mapinfo` has no method matching mapinfo(::Type{Images.ImageMagick}, ::Images.ImageCmap{Color.RGB{T<:Union{AbstractFloat,FixedPointNumbers.FixedPoint}},2,Array{UInt8,2}})
Closest candidates are:
  mapinfo(::Type{Color.RGB{FixedPointNumbers.UfixedBase{UInt8,8}}}, ::Any)
  mapinfo(::Type{Color.AlphaColorValue{Color.RGB{FixedPointNumbers.UfixedBase{UInt8,8}},FixedPointNumbers.UfixedBase{UInt8,8}}}, ::Any)
  mapinfo{CT<:Union{Color.AbstractAlphaColorValue{C<:Color.ColorValue{T},T<:Real},Color.ColorValue{T}}}(::Type{Color.RGB24}, ::AbstractArray{CT<:Union{Color.AbstractAlphaColorValue{C<:Color.ColorValue{T},T<:Real},Color.ColorValue{T}},N})
  ...
 in imwrite at /home/lobi/.julia/v0.4/Images/src/io.jl:189
 in include at ./boot.jl:254
 in include_from_node1 at ./loading.jl:264
while loading /home/lobi/juliarepo/test_image_cmap.jl, in expression starting on line 18

```

The first block shows an unstable Images.jl. I pulled from master (as Pkg.status()) shows and get now the 'map' problem. Also with other file formats.

**Comment by timholy** [comment]
Hmm. Try commenting out the `precompile` line of Images.jl (the very first line) and doing `rm ~/.julia/lib/v0.4/Images.ji`. That will disable precompilation. Which would be sad, but segfaults are worse.

**Comment by timholy** [comment]
P.S.: You should _not_ see the line
```
INFO: Recompiling stale cache file /home/lobi/.julia/lib/v0.4/Images.ji for module Images.
```
If you do, it means it's still using precompilation somehow (e.g., it's finding the `.ji` file somewhere else).

**Comment by timholy** [comment]
But see https://github.com/JuliaLang/julia/issues/12638. Maybe just upgrading julia will fix it?

**Comment by lobingera** [comment]
Sorry for mixing this. I think the main problem in this issue is: 
```
ERROR: LoadError: MethodError: `mapinfo` has no method matching mapinfo(::Type{Images.ImageMagick}, ::Images.ImageCmap{Color.RGB{T<:Union{AbstractFloat,FixedPointNumbers.FixedPoint}},2,Array{UInt8,2}})
```

The precompilation thing is not nice in testing, but worked on someplace else.

**Comment by timholy** [comment]
I had "fixed" the type instability in `cmaprgb = Array(RGB, 255)`, and that's why it passed my test but not yours. In general you should use `cmaprgb = Array(RGB{Float64}, 255)` (or some concrete type).

Should be fixed by #339 when that merges.
